File: native-ios/NaviLive/Sources/Services/Navigation/SearchResultCore.swift
import Foundation

enum SearchPlaceKind: Equatable {
  case shop
  case parcelLocker
  case railStation
  case busStop
  case tramStop
  case other
}

struct SearchRankingValues {
  let score: Int
  let distanceMeters: Int
  let importance: Double
  let isNearbyCandidate: Bool
}

/// Czyste reguły wyszukiwania używane przez adapter sieciowy i testy.
enum SearchResultCore {
  static func kindFromNominatim(category: String, type: String) -> SearchPlaceKind {
    if category == "shop" { return .shop }
    if category == "amenity", type == "parcel_locker" { return .parcelLocker }
    if category == "railway", ["station", "halt"].contains(type) { return .railStation }
    if category == "public_transport", type == "station" { return .railStation }
    if category == "highway", type == "bus_stop" { return .busStop }
    if category == "public_transport", ["platform", "stop_position"].contains(type) { return .busStop }
    if category == "railway", type == "tram_stop" { return .tramStop }
    return .other
  }

  /// PKP nie powinno zamieniać się w przystanek autobusowy lub tramwajowy.
  static func shouldKeepForRailQuery(kind: SearchPlaceKind, wantsTransitStop: Bool) -> Bool {
    wantsTransitStop || kind == .railStation || kind == .other
  }

  static func localNameMatches(normalizedName: String, nameSearchTerms: [String]) -> Bool {
    nameSearchTerms.isEmpty || nameSearchTerms.allSatisfy { normalizedName.contains($0) }
  }

  /// Porównuje kandydatów w kolejności prezentowania na liście.
  static func compareForDisplay(_ left: SearchRankingValues, _ right: SearchRankingValues) -> Bool {
    let leftDistance = sortableDistance(left.distanceMeters)
    let rightDistance = sortableDistance(right.distanceMeters)
    let scoreDifference = left.score - right.score
    if abs(scoreDifference) <= SharedProductRules.Search.nearbyBonus, leftDistance != rightDistance {
      return leftDistance < rightDistance
    }
    if scoreDifference != 0 { return scoreDifference > 0 }
    if left.isNearbyCandidate != right.isNearbyCandidate {
      return left.isNearbyCandidate && !right.isNearbyCandidate
    }
    if leftDistance != rightDistance { return leftDistance < rightDistance }
    return left.importance > right.importance
  }

  /// Wybiera lepszą wersję tego samego punktu zwróconą przez inne źródło.
  static func isBetterDuplicate(_ left: SearchRankingValues, _ right: SearchRankingValues) -> Bool {
    if left.score != right.score { return left.score > right.score }
    if left.isNearbyCandidate != right.isNearbyCandidate {
      return left.isNearbyCandidate && !right.isNearbyCandidate
    }
    if left.distanceMeters != right.distanceMeters {
      return sortableDistance(left.distanceMeters) < sortableDistance(right.distanceMeters)
    }
    return left.importance > right.importance
  }

  private static func sortableDistance(_ distanceMeters: Int) -> Int {
    distanceMeters > 0 ? distanceMeters : Int.max
  }
}

File: android/app/src/main/java/com/navilive/android/data/routing/SearchResultCore.kt
package com.navilive.android.data.routing

import com.navilive.android.model.SharedProductRules
import kotlin.math.abs

/** Kategorie używane do filtrowania i porządkowania wyników wyszukiwania. */
internal enum class SearchPlaceKind {
    Shop,
    ParcelLocker,
    RailStation,
    BusStop,
    TramStop,
    Other,
}

internal data class SearchRankingValues(
    val score: Int,
    val distanceMeters: Int,
    val importance: Double,
    val isNearbyCandidate: Boolean,
)

/** Czyste reguły wyszukiwania współdzielone przez adapter i testy jednostkowe. */
internal object SearchResultCore {

    fun kindFromNominatim(category: String, type: String): SearchPlaceKind {
        return when {
            category == "shop" -> SearchPlaceKind.Shop
            category == "amenity" && type == "parcel_locker" -> SearchPlaceKind.ParcelLocker
            category == "railway" && type in setOf("station", "halt") -> SearchPlaceKind.RailStation
            category == "public_transport" && type == "station" -> SearchPlaceKind.RailStation
            category == "highway" && type == "bus_stop" -> SearchPlaceKind.BusStop
            category == "public_transport" && type in setOf("platform", "stop_position") -> SearchPlaceKind.BusStop
            category == "railway" && type == "tram_stop" -> SearchPlaceKind.TramStop
            else -> SearchPlaceKind.Other
        }
    }

    /** PKP nie powinno zamieniać się w przystanek autobusowy lub tramwajowy. */
    fun shouldKeepForRailQuery(kind: SearchPlaceKind, wantsTransitStop: Boolean): Boolean {
        return wantsTransitStop || kind == SearchPlaceKind.RailStation || kind == SearchPlaceKind.Other
    }

    fun localNameMatches(normalizedName: String, nameSearchTerms: List<String>): Boolean {
        return nameSearchTerms.isEmpty() || nameSearchTerms.all { normalizedName.contains(it) }
    }

    /** Porównuje kandydatów w kolejności prezentowania na liście. */
    fun compareForDisplay(left: SearchRankingValues, right: SearchRankingValues): Int {
        val leftDistance = sortableDistance(left.distanceMeters)
        val rightDistance = sortableDistance(right.distanceMeters)
        val scoreDifference = left.score - right.score
        if (abs(scoreDifference) <= SharedProductRules.Search.nearbyBonus && leftDistance != rightDistance) {
            return leftDistance.compareTo(rightDistance)
        }
        if (scoreDifference != 0) return -scoreDifference
        if (left.isNearbyCandidate != right.isNearbyCandidate) {
            return if (left.isNearbyCandidate) -1 else 1
        }
        if (leftDistance != rightDistance) return leftDistance.compareTo(rightDistance)
        return -left.importance.compareTo(right.importance)
    }

    /** Wybiera lepszą wersję tego samego punktu zwróconą przez inne źródło. */
    fun isBetterDuplicate(left: SearchRankingValues, right: SearchRankingValues): Boolean {
        return when {
            left.score != right.score -> left.score > right.score
            left.isNearbyCandidate != right.isNearbyCandidate -> left.isNearbyCandidate
            left.distanceMeters != right.distanceMeters -> {
                sortableDistance(left.distanceMeters) < sortableDistance(right.distanceMeters)
            }
            else -> left.importance > right.importance
        }
    }

    private fun sortableDistance(distanceMeters: Int): Int {
        return if (distanceMeters > 0) distanceMeters else Int.MAX_VALUE
    }
}
